editor/src/clj/internal/graph.clj

```diff
@@ -829,58 +829,57 @@
-                                                                   compute-cached-outputs-function)]
-                                         (when (.contains cached-outputs output)
-                                           (.putIfAbsent cached-endpoints endpoint true)))
-                                       (.putIfAbsent cached-endpoints endpoint true))
-                                     (-> node-id
-                                         gt/node-id->graph-id
-                                         graph-id->node-successor-map
-                                         (get node-id)
-                                         (get output))))))
-                             endpoints)))
-        endpoints->tasks-xf (comp (partition-all 512) (map make-task!))
         cache-key (into [endpoints]
                         (map #(System/identityHashCode (val %)))
                         graph-id->node-successor-map)]
     (.get basis-dependencies-cache
           cache-key
           (reify Function
             (apply [_ _]
-              (loop [tasks (into [] endpoints->tasks-xf endpoints)]
-                (let [next-tasks
-                      (if (= 1 (count tasks))
-                        (into [] endpoints->tasks-xf ((nth tasks 0)))
-                        (into []
-                              (comp (mapcat deref) endpoints->tasks-xf)
-                              (.invokeAll pool tasks)))]
-                  (when (pos? (count next-tasks))
-                    (recur next-tasks))))
-              (.keySet cached-endpoints))))))
+              (let [pool (ForkJoinPool/commonPool)
+                    node-type->cached-outputs (ConcurrentHashMap.)
+                    all-endpoints (ConcurrentHashMap.)
+                    cached-endpoints (ConcurrentHashMap.)
+                    make-task! (fn [endpoints]
+                                 (fn []
+                                   (into []
+                                         (mapcat
+                                           (fn [endpoint]
+                                             (when (nil? (.putIfAbsent all-endpoints endpoint true))
+                                               (let [node-id (gt/endpoint-node-id endpoint)
+                                                     output (gt/endpoint-label endpoint)]
+                                                 (if-let [node (gt/node-by-id-at basis node-id)]
+                                                   (let [node-type (gt/node-type node)
+                                                         ^Set cached-outputs (.computeIfAbsent
+                                                                               node-type->cached-outputs
+                                                                               node-type
+                                                                               compute-cached-outputs-function)]
```

Comment:
Now that `node-type->cached-outputs` is recreated in the `basis-dependencies-cache` cache miss function every time, it could be beneficial to instead make `in/cached-outputs` memoized  on the `deref`:ed node-type so we don't have to redo the work on every `basis-dependencies-cache` miss.